v2.go

```diff
@@ -71,6 +75,107 @@ func handleAuthorizationSignInRequest(conn *websocket.Conn, action *Action) {
 	writeQueuedActions(client)
 	action.Type = authorizationSignInSuccess
 	client.writeJSON(action)
+
+	undeliveredMessages, err := findUndeliveredMessages(client.ID)
+	if err != nil {
+		log.Println("failed to find undelivered message ids")
+	}
+
+	for _, undeliveredMessage := range undeliveredMessages {
+		action := constructDeliverMessageAction(undeliveredMessage)
+		client.writeJSON(action)
+	}
+}
+
+func handleMessagingSendRequest(conn *websocket.Conn, action *Action) {
+	client, err := getSignedInClient(conn)
+	if err != nil {
+		client.writeFailure(messagingSendFailure, []string{err.Error()})
+		return
+	}
+
+	content, ok := action.Payload["content"].(string)
+	if !ok {
+		client.writeFailure(messagingSendFailure, []string{"content is missing"})
+		return
+	}
+
+	messageGroupID, ok := action.Payload["message_group_id"].(float64)
+	if !ok {
+		client.writeFailure(messagingSendFailure, []string{"message_group_id is missing"})
+		return
+	}
+
+	senderID, ok := action.Payload["sender_id"].(float64)
+	if !ok {
+		client.writeFailure(messagingSendFailure, []string{"sender_id is missing"})
+		return
+	}
+
+	if client.ID != int(senderID) {
+		client.writeFailure(messagingSendFailure, []string{"sender_id mismatch"})
+		return
+	}
+
+	if !isClientInMessageGroup(int(senderID), int(messageGroupID)) {
+		client.writeFailure(messagingSendFailure, []string{"sender doesn't belong to message group"})
+		return
+	}
+
+	message, err := createMessage(content, int(messageGroupID), int(senderID))
+	if err != nil {
+		client.writeFailure(messagingSendFailure, []string{err.Error()})
+		return
+	}
+
+	messageGroupMemberIDs, err := findAllMessageGroupMemberIDs(message.MessageGroupID)
+	if err != nil {
+		client.writeFailure(messagingSendFailure, []string{err.Error()})
+		return
+	}
+
+	var recipientIDs []int
+
+	for _, recipientID := range messageGroupMemberIDs {
+		if recipientID == int(message.SenderID) {
+			continue
+		}
+
+		recipientIDs = append(recipientIDs, recipientID)
+	}
+
+	for _, recipientID := range recipientIDs {
+		createReceipt(message.ID, recipientID)
+	}
+
+	messageDeliveryChan <- &MessageDelivery{
+		Message:      message,
+		RecipientIDs: recipientIDs,
+	}
+
+	payload := make(map[string]interface{})
+	payload["message_id"] = message.ID
+	client.writeJSON(&Action{
+		Payload: payload,
+		Type:    messagingSendSuccess,
+	})
+}
+
+func handleMessagingDeliverSuccess(conn *websocket.Conn, action *Action) {
+	client, err := getSignedInClient(conn)
+	if err != nil {
+		client.writeFailure(messagingDeliverFailure, []string{err.Error()})
+		return
+	}
+
+	messageID, ok := action.Payload["message_id"].(float64)
+	if !ok {
+		client.writeFailure(messagingDeliverFailure, []string{"message_id is missing"})
+		return
+	}
+
+	recipientID := client.ID
+	updateReceiptDateDelivered(int(messageID), recipientID)
```

Comment:
@cglotr Error handling here?

Reply:
Done 5293223. But, I realized something. When the client send a `*_SUCCESS` to the server, the server might fail. Which means the server will send `*_FAILURE` to the client.

Eg:
1. Server sends `MESSAGING_DELIVER_REQUEST`
2. Client responds with `MESSAGING_DELIVER_SUCCESS`
3. Server fails and sends `MESSAGING_DELIVER_FAILURE`